asgineer/utils.py

```diff
@@ -138,12 +137,17 @@ async def asset_handler(request, path=None):
             headers["content-encoding"] = "gzip"
             body = zipped[path]
         else:
-            body = assets[path]
+            body = to_bytes(assets[path])
+
+        headers["content-length"] = str(len(body))
```

Comment:
It was required before I made the change to `utils.py`; I'll see if that's no longer the case.

Reply:
OK, I think that will work. I'll update the PR.

Btw, here are the relevant parts of the relevant RFCs.

> ### [3.3.2.  Content-Length](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.2)
>
> ...
>
> A server MAY send a Content-Length header field in a 304 (Not
   Modified) response to a conditional GET request (Section 4.1 of
   [RFC7232]); a server MUST NOT send Content-Length in such a response
   unless its field-value equals the decimal number of octets that would
   have been sent in the payload body of a 200 (OK) response to the same
   request.
>
> ...

> ## [4.1.  304 Not Modified](https://www.rfc-editor.org/rfc/rfc7232#section-4.1)
>
>   The 304 (Not Modified) status code indicates that a conditional GET
   or HEAD request has been received and would have resulted in a 200
   (OK) response if it were not for the fact that the condition
   evaluated to false.  In other words, there is no need for the server
   to transfer a representation of the target resource because the
   request indicates that the client, which made the request
   conditional, already has a valid representation; the server is
   therefore redirecting the client to make use of that stored
   representation as if it were the payload of a 200 (OK) response.
>
>   The server generating a 304 response MUST generate any of the
   following header fields that would have been sent in a 200 (OK)
   response to the same request: Cache-Control, Content-Location, Date,
   ETag, Expires, and Vary.
>
>   Since the goal of a 304 response is to minimize information transfer
   when the recipient already has one or more cached representations, a
   sender SHOULD NOT generate representation metadata other than the
   above listed fields unless said metadata exists for the purpose of
   guiding cache updates (e.g., Last-Modified might be useful if the
   response does not have an ETag field).
>
>   Requirements on a cache that receives a 304 response are defined in
   Section 4.3.4 of [RFC7234].  If the conditional request originated
   with an outbound client, such as a user agent with its own cache
   sending a conditional GET to a shared proxy, then the proxy SHOULD
   forward the 304 response to that client.
>
>   A 304 response cannot contain a message-body; it is always terminated
   by the first empty line after the header fields.

I'll revise accordingly.

Reply:
OK, this was a bigger change. My intention was to make the flow more top-to-bottom, and the decisions — i.e., removing certain headers from 304 responses — more obvious. If the changes make you sad, let me know. :-)

## Localhost

```bash
# 307 Response

$ curl -v "http://localhost:5555/" -o /dev/null
...
< HTTP/1.1 307 Temporary Redirect
< date: Fri, 08 Mar 2024 18:55:59 GMT
< server: uvicorn
< Location: /timetagger/
< Transfer-Encoding: chunked

# 200 Response

$ curl -v "http://localhost:5555/timetagger/" -o /dev/null
...
< HTTP/1.1 200 OK
< date: Fri, 08 Mar 2024 18:56:10 GMT
< server: uvicorn
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9196
...

# 200 Response (HEAD Request)

$ curl -v -I "http://localhost:5555/timetagger/" -o /dev/null
 ...
< HTTP/1.1 200 OK
< date: Fri, 08 Mar 2024 18:56:26 GMT
< server: uvicorn
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9196
...

# 304 Response

$ curl -v -H "If-None-Match: \"512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349\"" "http://localhost:5555/timetagger/" -o /dev/null
...
< HTTP/1.1 304 Not Modified
< date: Fri, 08 Mar 2024 18:56:37 GMT
< server: uvicorn
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< cache-control: public, must-revalidate, max-age=0
...
```

## Server (behind Apache)

```bash
# 307 Response

$ curl -v "https://timetaggerpatched.nfshost.com/" -o /dev/null
 ...
< HTTP/2 307 
< date: Fri, 08 Mar 2024 19:00:30 GMT
< server: Apache
< location: /timetagger/
< vary: Accept-Encoding
< via: e1s
< content-length: 0
 ...

# 200 Response

$ curl -v "https://timetaggerpatched.nfshost.com/timetagger/" -o /dev/null
 ...
< HTTP/2 200 
< content-length: 9196
< date: Fri, 08 Mar 2024 19:00:39 GMT
< server: Apache
< cache-control: public, must-revalidate, max-age=0
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< vary: Accept-Encoding
< via: e1s
< content-type: text/html
 ...

# 200 Response (HEAD Request)

$ curl -v -I "https://timetaggerpatched.nfshost.com/timetagger/" -o /dev/null
 ...
< HTTP/2 200 
< content-length: 9196
< date: Fri, 08 Mar 2024 19:00:50 GMT
< server: Apache
< cache-control: public, must-revalidate, max-age=0
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< vary: Accept-Encoding
< via: e1s
< content-type: text/html
 ...

# 304 Response

$ curl -v -H "If-None-Match: \"512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349\"" "https://timetaggerpatched.nfshost.com/timetagger/" -o /dev/null
 ...
< HTTP/2 304 
< date: Fri, 08 Mar 2024 19:01:01 GMT
< server: Apache
< cache-control: public, must-revalidate, max-age=0
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< vary: Accept-Encoding
< via: e1s
...
```